pallets/randomness/src/types.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	traits::GetBabeData, BalanceOf, Config, Error, Event, Pallet, RandomnessResults, RequestId,
+};
+use frame_support::pallet_prelude::*;
+use frame_support::traits::{Currency, ExistenceRequirement::KeepAlive, ReservableCurrency};
+use pallet_evm::AddressMapping;
+use sp_core::{H160, H256};
+use sp_runtime::traits::{CheckedAdd, CheckedSub, Saturating};
+
+#[derive(PartialEq, Copy, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+/// Type of request
+/// Represents a request for the most recent randomness of this type at or after the inner time
+pub enum RequestType<T: Config> {
+	/// Babe per relay chain block
+	BabeCurrentBlock(T::BlockNumber),
+	/// Babe one epoch ago
+	BabeOneEpochAgo(u64),
+	/// Babe two epochs ago
+	BabeTwoEpochsAgo(u64),
+	/// Local per parachain block VRF output
+	Local(T::BlockNumber),
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// Raw randomness snapshot, the unique value for a `RequestType` in `RandomnessResults` map
+pub struct RandomnessResult<Hash> {
+	/// Randomness once available
+	pub randomness: Option<Hash>,
+	/// Number of randomness requests for the type
+	pub request_count: u64,
+}
+
+impl<Hash: Clone> RandomnessResult<Hash> {
+	pub fn new() -> RandomnessResult<Hash> {
+		RandomnessResult {
+			randomness: None,
+			request_count: 1u64,
+		}
+	}
+	/// Increment request count
+	pub fn increment_request_count<T: Config>(&self) -> Result<Self, DispatchError> {
+		let new_request_count = self
+			.request_count
+			.checked_add(1u64)
+			.ok_or(Error::<T>::RequestCounterOverflowed)?;
+		Ok(RandomnessResult {
+			randomness: self.randomness.clone(),
+			request_count: new_request_count,
+		})
+	}
+	/// Returns whether successfully decremented
+	/// Failure implies the randomness result should be removed from storage
+	pub fn decrement_request_count(&self) -> Option<Self> {
```

Comment:
It would be nice to have symmetry with the `increment` version of the fn, maybe:

* Return `DispatchResult`
* Underflow -> Err
* Any other value -> Ok
* Expect count of 0 to mean "dropable" rather than None

Reply:
I made it this way so that we drop the result when it goes to 0. This is how it is called:
```rust
if let Some(result) = RandomnessResults::<T>::take(&request.request.info) {
    if let Some(new_result) = result.decrement_request_count() {
	    RandomnessResults::<T>::insert(&request.request.info, new_result);
    } // else RandomnessResult is removed from storage
}
```